proteus/MeshAdaptPUMI/MeshConverter.cpp

```diff
@@ -420,3 +430,687 @@ int MeshAdaptPUMIDrvr::updateMaterialArrays(Mesh& mesh,
+    int gid = it->first;
+    int to = std::min(peers - 1, gid / quotient);
+    PCU_COMM_PACK(to, gid);
+  }
+  PCU_Comm_Send();
+  int myOffset = self * quotient;
+  /* brokers store all the part ids that sent messages
+     for each global id */
+  while (PCU_Comm_Receive()) {
+    int gid;
+    PCU_COMM_UNPACK(gid);
+    int from = PCU_Comm_Sender();
+    tmpParts.at(gid - myOffset).push_back(from);
+  }
+  /* for each global id, send all associated part ids
+     to all associated parts */
+  PCU_Comm_Begin();
+  for (int i = 0; i < mySize; ++i) {
+    std::vector<int>& parts = tmpParts[i];
+    for (size_t j = 0; j < parts.size(); ++j) {
+      int to = parts[j];
+      int gid = i + myOffset;
+      int nparts = parts.size();
+      PCU_COMM_PACK(to, gid);
+      PCU_COMM_PACK(to, nparts);
+      for (size_t k = 0; k < parts.size(); ++k)
+        PCU_COMM_PACK(to, parts[k]);
+    }
+  }
+  PCU_Comm_Send();
+  /* receiving a global id and associated parts,
+     lookup the vertex and classify it on the partition
+     model entity for that set of parts */
+  while (PCU_Comm_Receive()) {
+    int gid;
+    PCU_COMM_UNPACK(gid);
+    int nparts;
+    PCU_COMM_UNPACK(nparts);
+    Parts residence;
+    for (int i = 0; i < nparts; ++i) {
+      int part;
+      PCU_COMM_UNPACK(part);
+      residence.insert(part);
+    }
+    MeshEntity* vert = globalToVert[gid];
+    m->setResidence(vert, residence);
+  }
+}
+
+/* given correct residence from the above algorithm,
+   negotiate remote copies by exchanging (gid,pointer)
+   pairs with parts in the residence of the vertex */
+static void constructRemotes(Mesh2* m, GlobalToVert& globalToVert)
+{
+  int self = PCU_Comm_Self();
+  PCU_Comm_Begin();
+  APF_ITERATE(GlobalToVert, globalToVert, it) {
+    int gid = it->first;
+    MeshEntity* vert = it->second;
+    Parts residence;
+    m->getResidence(vert, residence);
+    APF_ITERATE(Parts, residence, rit)
+      if (*rit != self) {
+        PCU_COMM_PACK(*rit, gid);
+        PCU_COMM_PACK(*rit, vert);
+      }
+  }
+  PCU_Comm_Send();
+  while (PCU_Comm_Receive()) {
+    int gid;
+    PCU_COMM_UNPACK(gid);
+    MeshEntity* remote;
+    PCU_COMM_UNPACK(remote);
+    int from = PCU_Comm_Sender();
+    MeshEntity* vert = globalToVert[gid];
+    m->addRemote(vert, from, remote);
+  }
+}
+
+void construct(Mesh2* m, const int* conn, const int* conn_b, int nelem, 
+    int nelem_b, int nverts,int etype, int etype_b, int* local2globalMap,
+    GlobalToVert& globalToVert)
+{
+  constructVerts(m, nverts,local2globalMap,globalToVert);
+  constructBoundaryElements(m, conn_b, nelem_b, etype_b, globalToVert);
+  constructElements(m, conn, nelem, etype, globalToVert);
+  constructResidence(m, globalToVert);
+  constructRemotes(m, globalToVert);
+  stitchMesh(m);
+  m->acceptChanges();
+}
+
+}
+
+/**************************************************************************/
+
+
+//The following functions are used to facilitate and perform a reconstruction
```

Comment:
I'll start the issue. I can probably get something working in serial in a few days, but I expect some complications in parallel which may take a week, so let's say in a week and a half I'll aim to get something working in 3D parallel.